Dockerized benchmarking framework

oschaaf opened the pull request:
This PR has changes that make it easy to consume our integration testing as a benchmarking
tool framework for external repositories, allowing those to write custom tests for running in CI.

This probably needs to be split up into multiple PR's, but this allows reviewers to
have an early peek at what such PR's would lead up to.

The `README.md` in here has more context, but the primary goals are te be able to:
- run the tests against arbitrary Envoy revisions
- persist profile dumps, flamegraphs, and latency numbers per test
- run the nighthawk tools via docker
- offer stock tests, but also allow scaffolding consumer-specific tests

Example of a fully dockerized flow (more example in `README.md`):

```
# This script runs the dockerized benchmarking framework, which in
# turn will pull Nighthawk and Envoy in via docker. 

set -eo pipefail
set +x
set -u

# The benchmark logs and artifacts will be dropped here
OUTDIR="/home/oschaaf/code/envoy-perf-vscode/nighthawk/benchmarks/tmp/"
# Used to map the test that we want to see executed into the docker container
TEST_DIR="/home/oschaaf/code/envoy-perf-vscode/nighthawk/benchmarks/test/"

# Rebuild the docker in case something changed.
./docker_build.sh && 
docker run -it --rm \
  -v "/var/run/docker.sock:/var/run/docker.sock:rw" \
  -v "${OUTDIR}:${OUTDIR}:rw" \
  -v "${TEST_DIR}:/usr/local/bin/benchmarks/benchmarks.runfiles/nighthawk/benchmarks/external_tests/" \
  --network=host \
  --env NH_DOCKER_IMAGE="envoyproxy/nighthawk-dev:latest" \
  --env ENVOY_DOCKER_IMAGE_TO_TEST="envoyproxy/envoy-dev:f61b096f6a2dd3a9c74b9a9369a6ea398dbe1f0f" \
  --env TMPDIR="${OUTDIR}" \
  oschaaf/benchmark-dev:latest ./benchmarks --log-cli-level=info -vvvv 
```

Part of #245

oschaaf commented:
/cc @marcomagdy @antoniovicente

htuch commented on benchmarks/BUILD:
```diff
@@ -1,20 +1,33 @@
+        "test/test_discovery.py",
+    ],
+    data = [
+        "configurations/envoy_proxy.yaml",
```
Is it possible to depend on `:benchmarks` directly here to avoid having to redeclare things?

oschaaf replied:
I improved this, but could not eliminate all of the duplication, ptal

htuch commented on benchmarks/README.md:
```diff
@@ -0,0 +1,104 @@
+
+## Immediate goals
+
+The goal is te be able to:
```
Nit: to be able to

htuch commented on benchmarks/README.md:
```diff
@@ -0,0 +1,104 @@
+
+This scripts shows how to use the benchmarking suite.
+It will run a selection of an example [benchmarks](test/test_discovery.py)
+scavenged from `/benchmarks`, which injects Envoy between the 
```
Maybe say "extracted" or "sourced" rather than "scavenged" :)

htuch commented on benchmarks/README.md:
```diff
@@ -0,0 +1,104 @@
+- run the suite against arbitrary Envoy revisions
+- persist profile dumps, flamegraphs, and latency numbers per test
+- run the nighthawk tools via docker
+- offer stock tests, but also allow scaffolding consumer-specific tests
```
Maybe point out that this is also intended for the use in A/B testing by developers and integration with GitHub CI flows (via the Docker support).

htuch commented on benchmarks/README.md:
```diff
@@ -0,0 +1,104 @@
+
+## Example: running with binaries
+
+This will build the Nighthawk binaries from the c++ code, and use those to
```
Nit: s/c++/C++/g

htuch commented on benchmarks/README.md:
```diff
@@ -0,0 +1,104 @@
+The goal is te be able to:
+- run the suite against arbitrary Envoy revisions
+- persist profile dumps, flamegraphs, and latency numbers per test
+- run the nighthawk tools via docker
```
Nit: s/docker/Docker/g

htuch commented on benchmarks/README.md:
```diff
@@ -0,0 +1,104 @@
+## Example: fully dockerized flow
+
+The framework can be run via docker and used that way to execute
+python benchmarks scripts not sourced from this repository, but
```
Nit: s/python/Python/g

htuch commented on benchmarks/README.md:
```diff
@@ -0,0 +1,104 @@
+  -v "${TEST_DIR}:/usr/local/bin/benchmarks/benchmarks.runfiles/nighthawk/benchmarks/external_tests/" \
+  --network=host \
+  --env NH_DOCKER_IMAGE="envoyproxy/nighthawk-dev:latest" \
+  --env ENVOY_DOCKER_IMAGE_TO_TEST="envoyproxy/envoy-dev:f61b096f6a2dd3a9c74b9a9369a6ea398dbe1f0f" \
```
How do you specify a custom Envoy binary in the Docker flow?

oschaaf replied:
This is documented in the flow with binaries. You don't specify a docker image, but instead just use `ENVOY_PATH` to point out your Envoy binary.

oschaaf replied:
That certainly is possible, but it result in a large image (I assume the binaries would need to be unstripped and build with all the right flags for profiling?) .

htuch replied:
Yeah. I don't think it needs to be in this PR, but it's a pretty useful feature IMHO. One thing that I wanted to do is run the exact same Docker image on my local workstation and different sized Cloud VMs to see what performance range is observed.

oschaaf replied:
Added this to the list of things that need to be done in `README.md`

htuch commented on benchmarks/README.md:
```diff
@@ -0,0 +1,104 @@
+
+# FUTURE
+
+- profiling / flamegraphing via perf/bcc tools; include the proxy-wasm flamegraphing research r&d
```
I think the most useful things to add are (in my $0.02 stack rank):
1. The ability to repeat the runs multiple times and obtain stats, e.g. how much variance there is, mean, etc.
2. The ability to obtain flamegraphs.
3. The ability to do A/B testing, similar to https://github.com/envoyproxy/envoy-perf/blob/master/siege/siege.py#L3.

But, I'm sure there are other folks who have other priorities here ;-)

oschaaf replied:
Also moved these into the future list in README.md

htuch commented on benchmarks/configurations/envoy_proxy.yaml:
```diff
@@ -0,0 +1,44 @@
+    hosts:
+    - socket_address:
+        address: $server_ip
+        port_value: $server_port
```
In addition to A/B tests, perhaps we could have A/B/C at some point, where A/B are the before/after Envoy binaries corresponding to some change and C is direct NH client to test server, to demonstrate the available source/sink capacity.

oschaaf replied:
I think if we would have a way to annotate tests with group metadata, then we can go all out here in another layer of python scripting, which performs post-processing based on that metadata. E.g. we can treat a group of tests annotated with the same tag as a/b/c/... and crunch the numbers + visualize like that.

htuch commented on benchmarks/docker/docker_build.sh:
```diff
@@ -0,0 +1,39 @@
+DOCKER_IMAGE_PREFIX="oschaaf/${DOCKER_NAME}"
+BAZEL_BIN="$(bazel info bazel-bin)"
+WORKSPACE="$(bazel info workspace)"
+bazel build //benchmarks:benchmarks
```
Why not use `BAZEL_BIN` here?

oschaaf replied:
I'm not sure what you mean here, could you elaborate?

htuch replied:
I think I was confused by `BAZEL_BIN` above. This is the `bazel-bin` dir, not the bazel binary..

htuch commented on benchmarks/docker/docker_build.sh:
```diff
@@ -0,0 +1,39 @@
+# consuming script that is off guard.
+
+DOCKER_NAME="benchmark"
+DOCKER_IMAGE_PREFIX="oschaaf/${DOCKER_NAME}"
```
Do you want to use `${USER}` here instead of `oschaaf`?

htuch commented on benchmarks/docker/docker_build.sh:
```diff
@@ -0,0 +1,39 @@
+BAZEL_BIN="$(bazel info bazel-bin)"
+WORKSPACE="$(bazel info workspace)"
+bazel build //benchmarks:benchmarks
+TMP_DIR="${WORKSPACE}/tmp-docker-build-context"
```
Why not use `mkdtemp`? Then you don't have to `rm -rf` in the next line.

oschaaf replied:
I think `mkdtemp` is a python thing, whereas this is bash? At least my system can't find that binary :-)

htuch commented on benchmarks/docker/test.sh:
```diff
@@ -0,0 +1,25 @@
+  -v "${TEST_DIR}:/usr/local/bin/benchmarks/benchmarks.runfiles/nighthawk/benchmarks/external_tests/" \
+  --network=host \
+  --env NH_DOCKER_IMAGE="envoyproxy/nighthawk-dev:latest" \
+  --env ENVOY_DOCKER_IMAGE_TO_TEST="envoyproxy/envoy-dev:f61b096f6a2dd3a9c74b9a9369a6ea398dbe1f0f" \
```
When and who will update this SHA?

oschaaf replied:
This is actually an accidental test script that got committed. Maybe it can serve a purpose though, as a NH CI task which ought to succeed. The Envoy sha could then be sourced from our dependency or some such. Regardless it should at least be a variable, maybe defaulting to `latest`

htuch replied:
I think drop it, but you can use fragments in a doc perhaps.

oschaaf replied:
I'll move this into README.md. In the future, we could move this and turn it into a ci task: we fetch a simple 101 test via http, and check that it works like we expect, to at least have some check in place this doesn't break.

htuch commented on benchmarks/test/test_discovery.py:
```diff
@@ -0,0 +1,90 @@
+from envoy_proxy import (inject_envoy_http_proxy_fixture, proxy_config)
+
+
+def run_with_cpu_profiler(fixture,
```
Why is this in discovery? Isn't this just general execution ochestration?

oschaaf replied:
I may want to pull this out of this PR and host it as an example on a private repo or some such. WDYT?

htuch replied:
Refactoring makes sense, or are you saying the whole file is moved somewhere?

oschaaf replied:
I was suggesting the possibility of moving it out of here and hosting it somewhere else. But after thinking about it: let's keep it in here, I'll see if I can move it into a more sensible place then. We will eventually need a location that hosts "foreign" tests for testing the full docker flow though.

htuch commented on test/integration/integration_test.py:
```diff
@@ -23,6 +23,7 @@
           test_selection_arg,  # Passed in via BUILD/py_test()
           "-x",
           path,
+          "--log-cli-level=debug",
```
?

htuch commented on test/integration/test_integration_zipkin.py:
```diff
@@ -1,7 +1,7 @@
 import pytest
 
-from integration_test_fixtures import (http_test_server_fixture)
+from integration_test_fixtures import (http_test_server_fixture, server_config)
```
Doesn't look like `server_config` is used here?

oschaaf replied:
It doesn't look like, but `http_test_server_fixture` relies on it. If we don't explicitly import `server_config` here, we will get an error [1]. I think that is kind of silly, but it's also not
something I can do anything about, this seems to be related to something in the way that
pytest works. I've spend some time looking into this, and so far the easiest resolution seems to
be to just go with the flow and import the `server_config` fixture in scripts that use fixtures that rely on it. I'll add a comment about it for clarification.

[1]
```
  @pytest.fixture(params=determineIpVersionsFromEnvironment())
  def http_test_server_fixture(request, server_config):
E       fixture 'server_config' not found
>       available fixtures: capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, http_test_server_fixture, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, testrun_uid, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, worker_id
>       use 'pytest --fixtures [testpath]' for help on them.
```

htuch replied:
Ack, thanks for the explanation.

htuch commented on test/integration/test_remote_execution.py:
```diff
@@ -2,7 +2,7 @@
 import pytest
 
-from test.integration.integration_test_fixtures import http_test_server_fixture
+from test.integration.integration_test_fixtures import (http_test_server_fixture, server_config)
```
Ditto

htuch commented:
CC @adisuissa

htuch commented on benchmarks/docker/test.sh:
```diff
@@ -0,0 +1,25 @@
+set -u
+
+# The benchmark logs and artifacts will be dropped here
+OUTDIR="/home/oschaaf/code/envoy-perf-vscode/nighthawk/benchmarks/tmp/"
```
These are hardcoded paths, I think I may have said something about `$USER` in another thread.

htuch commented on test/integration/nighthawk_test_server.py:
```diff
@@ -30,41 +33,82 @@ def __init__(self, server_binary_path, config_template_path, server_ip, ip_versi
+    self.parameters["tmpdir"] = self.tmpdir
+    self.parameters["tag"] = tag
+
+    def substitute_yaml_values(obj, params):
```
Would it be possible to feature freeze and land the PR as is before adding new things like this? This will make it easier to land I reckon.

oschaaf replied:
Backed this out in favor of hard coding the cert in the configs.

htuch commented on test/integration/nighthawk_test_server.py:
```diff
@@ -30,41 +33,82 @@ def __init__(self, server_binary_path, config_template_path, server_ip, ip_versi
+      return obj
+
+    r = runfiles.Create()
+    with open(r.Rlocation(self.config_template_path)) as f:
```
Why not just make the entire config a giant `string.Template` or Python f-string? That way you don't need to do YAML traversal.

oschaaf replied:
Done

oschaaf commented:
@htuch added a benchmark test in CI for NH itself, artifacts, [example](https://app.circleci.com/pipelines/github/envoyproxy/nighthawk/1597/workflows/dc20d1ae-2cab-4fcc-8104-1add64c87091/jobs/12559/artifacts)

Right now the most interesting artifacts are probably the html plots, e.g.:
[direct](https://12559-180498819-gh.circle-artifacts.com/0/2020-06-05-11-47-18/http_h1_small_request_small_reply_direct_IpVersion.IPV4-nighthawk_test_integration_configurations_nighthawk_http_origin.yaml/simple_plot.html) or [via our test server as a proxy](
https://12559-180498819-gh.circle-artifacts.com/0/2020-06-05-11-47-18/http_h1_small_request_small_reply_via_IpVersion.IPV4-nighthawk_test_integration_configurations_nighthawk_http_origin.yaml-nighthawk_benchmarks_configurations_envoy_proxy.yaml/simple_plot.html)

The collected profiles from Envoy's built-in profiler are also in there, but these are probably not very useful at this time; one would need access to the binaries and libraries involved to be able to use them properly.

htuch commented:
@oschaaf thanks. I'm hoping we can stabilize and land what is in the PR right now so we don't have it grow too much more :)

htuch commented on benchmarks/run_benchmark.sh:
```diff
@@ -0,0 +1,36 @@
+
+function cleanup() {
+    if [[ "$OSTYPE" == "darwin"* ]]; then
+        docker rm -f $(docker ps -a -q) || true
```
This seems dangerous; what if other Docker images are running?

oschaaf replied:
Removed this. The OSX variant will need a little bit of love anyway it broke somewhere along the way. I can do that in another iteration - after we land this PR - if someone needs that.

htuch commented on benchmarks/run_benchmark.sh:
```diff
@@ -0,0 +1,36 @@
+export ENVOY_DOCKER_IMAGE_TO_TEST="envoyproxy/envoy-dev:latest"
+
+# run all tests starting with test_http_h1_small
+bazel-bin/benchmarks/benchmarks --log-cli-level=info -vvvv benchmarks/
```
Any concerns over the impact of the level of verbosity and benchmark performance?

oschaaf replied:
No the log verbosity won't impact benchmark performance. There's only some things pre and post execution that are getting logged.

htuch commented on benchmarks/test/test_discovery.py:
```diff
@@ -0,0 +1,103 @@
+    with open(os.path.join(fixture.test_server.tmpdir, "simple_plot.html"), "w") as w:
+      # This will source nighthawk.json over http from the same dir.
+      # TODO(oschaaf): consider injecting the json directly into a html script source.
+      w.write(str.join("", txt))
```
Are you just trying to copy the file here? Why not use direct utils for this such as `shutil.copyfile`?

oschaaf replied:
I anticipated we'd might want to run replacements here, but so far we don't, so yeah, followed up and using `shutil.copyfile` now.

htuch commented on ci/do_ci.sh:
```diff
@@ -90,6 +97,33 @@ function do_tsan() {
+    # was a bad choice.
+    # Consider using a different environment variable for the benchmark tooling
+    # to use for this.
+    rm -rf ${TMPDIR}/tmp.*
```
This seems very scary of `${TMPDIR}` isn't set. It might be a good idea to verify that first.

oschaaf replied:
I safeguarded this for a bit by ensuring tmpdir is > 4 characters and the directory exists, wdyt?

htuch replied:
SGTM

htuch commented on test/integration/configurations/nighthawk_https_origin.yaml:
```diff
@@ -1,40 +1,146 @@
+                  private_key:
+                    inline_string: "-----BEGIN RSA PRIVATE KEY-----
+
+                      MIIEpAIBAAKCAQEAuvPdQdmwZongPAgQho/Vipd3PZWrQ6BKxIb4l/RvqtVP321I
```
Generally not a fan of inlining these, since when the certs expire, it's hard to know how to regenerate them. All the test certs in the Envoy tree have associated scripts to rebuild them.

oschaaf replied:
So to make sure I get it right, you'd like to see https://github.com/envoyproxy/nighthawk/pull/337/commits/605e5ce39369a2ed93505fb8c0498849cf82bf70 back in, but then instead of parsing yaml we do a string based substitution?

htuch replied:
Yeah. My main issue with that PR was that it was doing sophisticated YAML parsing; I think that since YAML is a textual format, you can just substitute directly into it. If you use something like Python's string.Template format in the YAML, there's no need to even write a custom substituter.

oschaaf replied:
@htuch following our Slack discussion, I pushed 345972b

snowp commented on benchmarks/test/templates/simple_plot.html:
```diff
@@ -0,0 +1,150 @@
+                var statistic = findStatisticsById(result, stat);
+                var percentiles = statistic["percentiles"];
+                var points = getPoints(percentiles);
+                //var samples = statistic["count"]
```
remove?

oschaaf replied:
done

oschaaf commented on .circleci/config.yml:
```diff
@@ -1,5 +1,5 @@
 references:
-  envoy-build-image: &envoy-build-image
+  envoy-build-image: &envoy-build-image 
```
removed

oschaaf commented on benchmarks/envoy_proxy.py:
```diff
@@ -0,0 +1,89 @@
+  def setUp(self):
+    super(InjectHttpProxyIntegrationTestBase, self).setUp()
+    logging.info("injecting envoy proxy ...")
+    # TODO(oschaaf): how should this interact with multiple backends?
```
Great point; added this to the TODO list in `README.md`. This may or may not be a doc issue, possibly this can already be made to work with the status quo and just needs a clear example of how to do it. (The NH integration tests does have tests involving multiple origins).

oschaaf commented:
@snowp thanks for having a look, I addressed your comments
@htuch I addressed most of your comments, modulo the one about hard-coding ssl certs, I had a question there instead.

oschaaf commented:
Thanks for the review @htuch 
@mum4k are you comfortable with me merging this?

mum4k commented on benchmarks/envoy_proxy.py:
```diff
@@ -0,0 +1,89 @@
+  the docker image and binary name.
+  """
+
+  def __init__(self, config_template_path, server_ip, ip_version, parameters=dict(), tag=""):
```
Can we add a comment for the initializer describing its arguments?

The same comment applies to all public functions, classes and methods.

mum4k commented on benchmarks/envoy_proxy.py:
```diff
@@ -0,0 +1,89 @@
+  is implemented as a customized Envoy, which is convenient here: the CLI and admin interface
+  mechanics that we rely on are the same. So all we do here, is specialize so we can override
+  the docker image and binary name.
+  """
```
Can we also document any public attributes inside an "Attributes" section of the class docstring?

mum4k replied:
Would it make sense to add an example of use inside the class docstring?

mum4k commented on benchmarks/envoy_proxy.py:
```diff
@@ -0,0 +1,89 @@
+class InjectHttpProxyIntegrationTestBase(HttpIntegrationTestBase):
+  """
+  Fixture which spins up a Nighthawk test server which listens on plain http,
+  and set up an Envoy instances to proxy to that, als listening on plain http.
```
(nit) Sentence doesn't read well around "als listening".

mum4k commented on benchmarks/envoy_proxy.py:
```diff
@@ -0,0 +1,89 @@
+  def __init__(self, ip_version, server_config, proxy_config):
+    """See base class."""
+    super(InjectHttpProxyIntegrationTestBase, self).__init__(ip_version, server_config)
+    self.proxy_config = proxy_config
```
Similarly in this class, let's document any public attributes. Alternatively we could make them private if they don't have to be public.

self._proxy_config

mum4k commented on benchmarks/envoy_proxy.py:
```diff
@@ -0,0 +1,89 @@
+
+  def getTestServerRootUri(self):
+    """See base class."""
+    r = super(InjectHttpProxyIntegrationTestBase, self).getTestServerRootUri()
```
Can we use a more descriptive name? E.g. "root_uri".

mum4k commented on benchmarks/envoy_proxy.py:
```diff
@@ -0,0 +1,89 @@
+@pytest.fixture(params=determineIpVersionsFromEnvironment())
+def inject_envoy_http_proxy_fixture(request, server_config, proxy_config):
+  '''
+  Injects an Envoy proxy.
```
Can we expand the doc comment by adding a section describing the arguments and the returned / yielded value?

mum4k commented on benchmarks/envoy_proxy.py:
```diff
@@ -0,0 +1,89 @@
+  '''
+  Injects an Envoy proxy.
+  '''
+  f = InjectHttpProxyIntegrationTestBase(request.param, server_config, proxy_config)
```
Similarly here, can we use a more descriptive name?

mum4k commented on benchmarks/test/templates/simple_plot.html:
```diff
@@ -0,0 +1,168 @@
+<head>
```
Can we document this file and the functions in it?

mum4k commented on benchmarks/test/templates/simple_plot.html:
```diff
@@ -0,0 +1,168 @@
+    }
+    function findWorkersResults(data) {
+      var results = data["results"];
+      var a = [];
```
Can we find a more descriptive name?

Applies to multiple locations.

mum4k commented on benchmarks/test/test_discovery.py:
```diff
@@ -0,0 +1,117 @@
+import os
+from test.integration.integration_test_fixtures import (http_test_server_fixture,
+                                                        https_test_server_fixture)
+from test.integration.utility import *
```
Can we import specific packages / modules?

oschaaf commented:
@mum4k thanks, I think I addressed everything. In addition, I added a small tweak to way docker gets run to avoid breakage in the docker flows when they use the latest Envoy revisions:
https://github.com/envoyproxy/nighthawk/pull/337/commits/c24062fb4f5f80896d2264696916b51cdb13f7fe#diff-281f9acd4aa73a1b04f644ae61f98febR94

oschaaf commented:
Another thing I should explicitly call out is that for `benchmark/` I added some more python flake8 checks in CI.
Generally check more things, and specifically add docstring (format) checks.

https://github.com/envoyproxy/nighthawk/pull/337/files#diff-ae86159524d9bc451d3e3cc04dc48bb2

mum4k commented on test/integration/integration_test_fixtures.py:
```diff
@@ -17,6 +17,8 @@
 from test.integration.nighthawk_test_server import NighthawkTestServer
 from test.integration.nighthawk_grpc_service import NighthawkGrpcService
 
+TIMESTAMP = time.strftime('%Y-%m-%d-%H-%M-%S')
```
(nit) Can we either make this private (i.e. _TIMESTAMP), or document it?

mum4k commented on test/integration/integration_test_fixtures.py:
```diff
@@ -45,24 +47,28 @@ class IntegrationTestBase():
   """
 
-  def __init__(self, ip_version, backend_count=1):
+  def __init__(self, ip_version, server_config, backend_count=1):
```
Can we document server_config in the Args: section of the docstring?

mum4k commented on test/integration/integration_test_fixtures.py:
```diff
@@ -45,24 +47,28 @@ class IntegrationTestBase():
-    self.nighthawk_test_server_path = os.path.join(self.test_rundir, "nighthawk_test_server")
-    self.nighthawk_test_config_path = None
-    self.nighthawk_client_path = os.path.join(self.test_rundir, "nighthawk_client")
+    self.confdir = "nighthawk/test/integration/configurations/"
```
We are adding some public attributes, can we document them in the class docstring? It will ideally have an "Attributes:" section.

This is partially pre-existing since none of the other attributes are documented there. Up to you if we document all now, or leave a todo for the pre-existing ones. We should still prefer making anything that can be private.

mum4k commented on test/integration/integration_test_fixtures.py:
```diff
@@ -84,15 +90,22 @@ def setUp(self):
-        "[", "_").replace("]", "")
-    self.parameters["test_id"] = test_id
+    if os.getenv("NH_DOCKER_IMAGE", "") == "":
+      assert (os.path.exists(self.nighthawk_test_server_path))
```
Can we amend these assertions with some error message that would help the receiver of this error message understand what wen't wrong?

Something like:
assert False, "custom error message"

mum4k commented on test/integration/integration_test_fixtures.py:
```diff
@@ -84,15 +90,22 @@ def setUp(self):
+      assert (os.path.exists(self.nighthawk_test_server_path))
+      assert (os.path.exists(self.nighthawk_client_path))
+
+    self.test_id = os.environ.get('PYTEST_CURRENT_TEST').split(':')[-1].split(' ')[0].replace(
```
Ideally attributes on classes are all initialized (and documented if public) inside the __init__ method. Crating attributes "dynamically" in other methods makes it harder to find the complete list (API) of the class.

Can we initialize self.test_id to some useful value inside __init__? If nothing else, we can set it to the zero value of the same type (an empty string?).

mum4k commented on test/integration/integration_test_fixtures.py:
```diff
@@ -84,15 +90,22 @@ def setUp(self):
-      test_server = NighthawkTestServer(self.nighthawk_test_server_path,
-                                        self.nighthawk_test_config_path, self.server_ip,
-                                        self.ip_version, self.parameters)
+      test_server = NighthawkTestServer(
```
This setUp method is getting fairly long. Any chance we can extracts some logical sub-parts from it and move it into well named private functions (or methods if necessary) ?

mum4k commented on test/integration/integration_test_fixtures.py:
```diff
@@ -214,26 +233,43 @@ def runNighthawkClient(self, args, expect_failure=False, timeout=30, as_json=Tru
       assert (client_process.returncode == 0)
     return output, logs
 
+  def transformNighthawkJson(self, json, format="human"):
```
Can we document this public method?

mum4k commented on test/integration/nighthawk_test_server.py:
```diff
@@ -30,41 +33,85 @@ def __init__(self, server_binary_path, config_template_path, server_ip, ip_versi
-      logging.info("Parameterized server configuration: %s", config)
-
-    with tempfile.NamedTemporaryFile(mode="w", delete=False, suffix=".yaml") as tmp:
+    self.docker_image = os.getenv("NH_DOCKER_IMAGE", "")
```
Note about documenting public attributes applies here as well.

mum4k commented on test/integration/nighthawk_test_server.py:
```diff
@@ -30,41 +33,85 @@ def __init__(self, server_binary_path, config_template_path, server_ip, ip_versi
+    self.parameters["tmpdir"] = self.tmpdir
+    self.parameters["tag"] = tag
+
+    def substitute_yaml_values(obj, params):
```
Can we define this as a standalone (global scope) function instead to make the __init__ method shorter and easier to follow?

Once we do, we should also document it.

mum4k commented on test/integration/nighthawk_test_server.py:
```diff
@@ -30,41 +33,85 @@ def __init__(self, server_binary_path, config_template_path, server_ip, ip_versi
+          if obj[0] == '$':
+            return Template(obj).substitute(params)
+          elif obj.startswith(INJECT_RUNFILE_MARKER):
+            r = runfiles.Create()
```
Here and below, can we use a better name than "r" ?

mum4k commented on test/integration/nighthawk_test_server.py:
```diff
@@ -30,41 +33,85 @@ def __init__(self, server_binary_path, config_template_path, server_ip, ip_versi
-    args = [
+    args = []
+    if self.docker_image != "":
+      # TODO(XXX): As of https://github.com/envoyproxy/envoy/commit/e8a2d1e24dc9a0da5273442204ec3cdfad1e7ca8
```
Did we mean to replace XXX with a value?

mum4k commented on test/integration/nighthawk_test_server.py:
```diff
@@ -30,41 +33,85 @@ def __init__(self, server_binary_path, config_template_path, server_ip, ip_versi
+
+    Path(self.tmpdir).mkdir(parents=True, exist_ok=True)
+
+    with tempfile.NamedTemporaryFile(
```
Are there any parts of this function that we could move out so that we make __init__ shorter and easier to follow?

mum4k commented on test/integration/nighthawk_test_server.py:
```diff
@@ -131,6 +179,18 @@ def __init__(self,
-                                              ip_version, "--config-path", parameters)
+                                              ip_version, "--config-path", parameters, tag)
+
+  def getCliVersionString(self):
```
Can we document this method?

mum4k commented on test/integration/test_grpc_service.py:
```diff
@@ -1,7 +1,7 @@
 import pytest
 
-from test.integration.integration_test_fixtures import http_test_server_fixture
+from test.integration.integration_test_fixtures import (http_test_server_fixture, server_config)
```
(optional for this PR) comment about only importing modules / packages applies here too.

Optional, because it is pre-existing and we probably don't want to increase the scope of this PR.

mum4k commented on test/integration/test_integration_basics.py:
```diff
@@ -30,7 +30,7 @@ def test_http_h1(http_test_server_fixture):
   assertCounterEqual(counters, "upstream_cx_total", 1)
   assertCounterEqual(counters, "upstream_cx_tx_bytes_total",
-                     1400 if http_test_server_fixture.ip_version == IpVersion.IPV6 else 1500)
+                     1400 if http_test_server_fixture.ip_version == IpVersion.IPV6 else 1450)
```
The two literals (1400, 1450) would benefit from being well named constants with some explanation.

oschaaf replied:
The change here is triggered by that we now connect to 0.0.0.0 instead of 127.0.0.1, resulting in a decrease of `upstream_cx_tx_bytes_total` as the request header gets slightly smaller.

oschaaf replied:
After discussing on Slack, we concluded this can stay as is; it's not as if this has anything to do with MTU sizes :-)

mum4k commented on test/integration/nighthawk_test_server.py:
```diff
@@ -18,61 +18,79 @@
   """
     Base class for running a server in a separate process.
+
+    Arguments:
```
(nit) The "Args:" section of a class docstring actually belongs to the docstring of the initializer (__init__). Would you mind moving it there?

I am also happy to do this in a follow-up PR since this is minor, just let me know.

oschaaf replied:
I prefer to do it in a follow up: this file exists out of `benchmarks/`, I'm hoping to tighten up the format checking so that I can get all the python to comply with the check format rules that are now specific to `benchmark/`. Being backed by that makes it easier, as the doc string format isn't in my muscle memory yet. Added a note to https://github.com/envoyproxy/nighthawk/issues/371 as a reminder.

oschaaf replied:
Forgot to mark this is ready for review after posting the comment above  - just did so.
